Dropbox 公布了 Atlas —— 一个托管服务编排平台 (#8261)
xitu/gold-miner: article/2021/dropbox-reveals-Atlas.md

> * 原文作者：[Eran Stiller](https://www.infoq.com/profile/Eran-Stiller/)
> * 译文出自：[掘金翻译计划](https://github.com/xitu/gold-miner)
> * 本文永久链接：[https://github.com/xitu/gold-miner/blob/master/article/2021/dropbox-reveals-Atlas.md](https://github.com/xitu/gold-miner/blob/master/article/2021/dropbox-reveals-Atlas.md)
> * 译者：[霜羽 Hoarfroster](https://github.com/PassionPenguin)

> > * 校对者：

Author: ```suggestion
> * 校对者：[5Reasons](https://github.com/5Reasons)、[kamly](https://github.com/kamly)
```

> > * 校对者：

Author: ```suggestion
> * 校对者：[5Reasons](https://github.com/5Reasons)、[kamly](https://github.com/kamly)、[husiyu](https://github.com/husiyu)
```

> > * 校对者：[5Reasons](https://github.com/5Reasons)、[kamly](https://github.com/kamly)

Author: ```suggestion
> * 校对者：[5Reasons](https://github.com/5Reasons)、[kamly](https://github.com/kamly)、[husiyu](https://github.com/husiyu)
```

> # Dropbox 公布了 Atlas —— 一个托管服务编排平台

Reviewer: Dropbox 发布托管服务编排平台 Atlas

> 在最近的博客文章中，[Dropbox 公布了 Atlas](https://dropbox.tech/infrastructure/atlas--our-journey-from-a-python-monolith-to-a-managed-platform)，旨在提供给用户面向服务的体系结构的各种好处，同时让拥有服务的运营成本降至最低。

Author: ```suggestion
在最近的博客文章中，[Dropbox 公布了 Atlas](https://dropbox.tech/infrastructure/atlas--our-journey-from-a-python-monolith-to-a-managed-platform)，旨在提供给用户面向服务的体系结构的各种好处，同时让拥有一个服务的运营成本降至最低。
```

> 在最近的博客文章中，[Dropbox 公布了 Atlas](https://dropbox.tech/infrastructure/atlas--our-journey-from-a-python-monolith-to-a-managed-platform)，该平台的目标是提供给用户[面向服务架构](https://en.wikipedia.org/wiki/Service-oriented_architecture)的各种好处，同时让拥有一个服务的运营成本降至最低。

Reviewer: 在最近的博客文章中，[Dropbox 发布了了 Atlas —— 一个托管服务编排平台](https://dropbox.tech/infrastructure/atlas--our-journey-from-a-python-monolith-to-a-managed-platform)，该平台的目标是给用户提供[面向服务架构](https://en.wikipedia.org/wiki/Service-oriented_architecture)的各种便利，同时让用户维护一个服务的运营成本降至最低。

Author: ```suggestion
在最近的博客文章中，[Dropbox 公布了 Atlas](https://dropbox.tech/infrastructure/atlas--our-journey-from-a-python-monolith-to-a-managed-platform)，该平台的目标是为用户提供[面向服务架构](https://en.wikipedia.org/wiki/Service-oriented_architecture)的各种便利，同时让拥有一个服务的维护成本降至最低。
```

> Atlas 的目标是支持小型的独立功能，为产品团队节省管理各种服务的开销，包括容量规划、警报设置等。Atlas 还借助后台自动调配的服务，为用户提供了与无服务器系统（如 AWS Fargate）搭配使用的体验。根据作者 Naphat Sanguansin 和 Utsav Shah 的说法，他们评估了使用现成的解决方案来运行该平台。但是，为了降低迁移风险并确保较低的工程成本，他们决定继续在 Dropbox 其余部分使用的同一编排平台上托管服务。

Author: ```suggestion
Atlas 的目标是支持小型的独立功能，为产品团队节省管理各种服务的开销，包括容量规划、警报设置等。Atlas 还借助后台自动调配的服务，为用户提供了与无服务器系统（如 AWS Fargate）搭配使用的体验。根据作者 Naphat Sanguansin 和 Utsav Shah 的说法，他们评估了使用现成的解决方案来运行该平台。但是，为了降低迁移风险并确保较低的工程成本，他们决定继续在 Dropbox 其余部分使用的同一部署安排平台上托管服务。
```

> Atlas 的目标是支持小型的独立功能，为产品团队节省管理各种服务的开销，包括容量规划、警报设置等。Atlas 还借助后台自动调配的服务，为用户提供了与无服务器系统（如 [AWS Fargate](https://aws.amazon.com/fargate/)），同时由后台自动配置的服务提供支持。根据作者 Naphat Sanguansin 和 Utsav Shah 的说法，他们评估了使用现成的解决方案来运行该平台。但是，为了降低迁移风险并确保较低的工程成本，他们决定继续在 Dropbox 其余部分使用的同一部署安排平台上托管服务。

Reviewer: ->> 但是，为了降低迁移风险并确保较低的工程成本，他们决定继续在与 Dropbox 其余部分相同的“部署-编排”平台上，托管这些服务。

Author: ```suggestion
Atlas 的目标是支持小型的独立功能，为产品团队节省管理各种服务的开销，包括容量规划、警报设置等。Atlas 还借助后台自动调配的服务，为用户提供了与无服务器系统（如 [AWS Fargate](https://aws.amazon.com/fargate/)），同时由后台自动配置的服务提供支持。根据作者 Naphat Sanguansin 和 Utsav Shah 的说法，他们评估了使用现成的解决方案来运行该平台。但是，为了降低迁移风险并确保较低的工程成本，他们决定继续在与 Dropbox 其余部分相同的“部署-编排”平台上，托管这些服务。
```

> 构造 Atlas 的原因是他们想要要替换 Dropbox 的 Python 中心库 [monolith](https://en.wikipedia.org/wiki/Monolith) Metaserver。构造 Altas 会是一个历时多年的历程，至今仍在进行之中。目前，Atlas 正在为它打算取代的 monolith 提供 25％ 以上的服务。作者给出了有关迁移过程的关键结论：

Author: ```suggestion
构建 Atlas 项目的原因是他们想要要替换 Dropbox 的 Python 中心库 [monolith](https://en.wikipedia.org/wiki/Monolith) Metaserver。Altas 的开发会是一个历时多年的历程，至今仍在进行之中。目前，Atlas 正在为它打算取代的 monolith 提供 25％ 以上的服务。作者给出了有关迁移过程的关键结论：
```

> > 多年努力中我们发现最重要的一点是，在项目生命周期的早期，编写经过深思熟虑的代码是至关重要的。否则，技术负担和代码复杂性将很快融合一起来作怪。导入周期的拆除和 Metaserver（...）的分解可能是该项目在战略上最有效的部分，因为它可以防止新代码导致问题并简化我们的代码。

Author: ```suggestion
> 多年努力中我们发现最重要的一点是，在项目生命周期的早期，编写经过深思熟虑的代码是至关重要的。否则，技术负担和代码复杂性将很快融合一起来作怪。导入周期的取消和 Metaserver（……）的分解可能是该项目在战略上最有效的部分，因为它可以防止新的代码导致问题，也同时能够简化我们的代码。
```

> > 多年努力中我们发现最重要的一点是，在项目生命周期的早期，编写经过深思熟虑的代码是至关重要的。否则，技术负担和代码复杂性将很快融合一起来作怪。导入周期的取消和 Metaserver（……）的分解可能是该项目在战略上最有效的部分，因为它可以防止新的代码导致问题，也同时能够简化我们的代码。

Reviewer: ->> 否则，技术负担和代码复杂性将很快一起来“作怪”。

Author: ```suggestion
> 多年努力中我们发现最重要的一点是，在项目生命周期的早期，编写经过深思熟虑的代码是至关重要的。否则，技术负担和代码复杂性会很快一起来作怪。导入周期的取消和 Metaserver（……）的分解可能是该项目在战略上最有效的部分，因为它可以防止新的代码导致问题，也同时能够简化我们的代码。
```

> 作者指出，由于代码库的大小和复杂性，以前许多改进 Metaserver 的努力都没有成功。这次，他们考虑[将其作为垫脚石，而不是里程碑](https://medium.com/@jamesacowling/stepping-stones-not-milestones-e6be0073563f)，为 Atlas 设计了执行计划。想法是，如果项目的下一部分由于任何原因失败，则每个增量步骤都将提供足够的价值。此策略的关键示例涉及对单片代码库进行改进，无论有没有 Atlas 实施，该改进都有其价值。此外，团队将为 Atlas 开发的许多增强功能回移植到 Metaserver 中，以进一步提高项目价值。

Author: ```suggestion
他们指出，出于代码库的大小和复杂性的问题，以前许多改进 Metaserver 的尝试都没有成功。这次，他们考虑[将 Altas 作一个垫脚石而不再是里程碑](https://medium.com/@jamesacowling/stepping-stones-not-milestones-e6be0073563f)，并为 Atlas 设计了执行计划。这个想法是不管项目的下一部分由于任何原因失败，每个逐步推进的步骤都已经有了它的价值。此策略的关键示例涉及对单片代码库进行改进，无论有没有 Atlas 实施，这个改进都有其价值。此外，开发团队会将为 Atlas 开发的许多增强功能回移植到 Metaserver 中，以进一步提高项目价值。
```

![https://res.infoq.com/news/2021/03/dropbox-atlas/en/resources/1Dropbox-atlas-before-after-1615307468409.png](https://res.infoq.com/news/2021/03/dropbox-atlas/en/resources/1Dropbox-atlas-before-after-1615307468409.png)
<small>以前与以后，图源 [Dropbox](https://dropbox.tech/infrastructure/atlas--our-journey-from-a-python-monolith-to-a-managed-platform) </small>

> Atlas 的设计涉及一些围绕组件化，编排和操作性的关键工作。Atlas 引入了 Atlasservlets（atlas servlets）作为 HTTP 路由的逻辑，原子分组，以改善组件化。作者说：“在为 Atlas 做准备时，我们与产品团队合作，将 Atlasservlet 分配给 Metaserver 中的每个路由，从而在 5000 多个路由中产生了 200 多个 Atlasservlet。” 每个 Servlet 均分配有一个所有者，该所有者是管理它的唯一权限。另外，要分解 Metaserver 代码库，他们必须打破我们的大多数 Python 导入周期。这个过程花了几年时间才能实现，。

Author: ```suggestion
Atlas 的设计涉及一些围绕组件化、安排和操作性的关键工作。Atlas 引入了 Atlasservlets（atlas servlets）作为 HTTP 路由的逻辑、最小化分组，以改善组件化。他们说：“在为 Atlas 做准备时，我们与产品团队合作，将 Atlasservlet 分配给 Metaserver 中的每个路由，从而在 5000 多个路由中构建了 200 多个 Atlasservlet。” 每个 Servlet 都分配有一个所有者 —— 拥有唯一的管理它的权限。另外，要分解 Metaserver 代码库，他们必须打破了大多数的 Python 导入周期。这个过程花了几年时间才能实现。
```

> 为了改进编排，Atlas 中的每个 Servlet 都是其自己的集群。默认情况下，该决定会提供隔离，因为行为异常的路由只会影响同一 Atlasservlet 中的其他路由。同样，此决定允许独立推送代码。此外，Dropbox 决定在 [gRPC](https://grpc.io/) 上进行标准化。继续为 HTTP 流量，他们用 [GRPC JSON 的转码器](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_json_transcoder_filter)中提供了 [Envoy](https://www.envoyproxy.io/)，这是他们在 Servlet 的前面作为代理服务器和负载平衡器使用。

Author: ```suggestion
为了改进安排，Atlas 中的每个 Servlet 都是其自己的集群。默认情况下这个思路隔离了它们，因为行为异常的路由应该只会影响同一 Atlasservlet 中的其他路由。同样，此决定允许独立去推送代码。此外，Dropbox 决定在 [gRPC](https://grpc.io/) 上进行标准化。为了继续处理 HTTP 流量，他们使用了 [GRPC JSON 转码器](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_json_transcoder_filter)中提供了的 [Envoy](https://www.envoyproxy.io/) —— 他们在 Servlet 的前面作为代理服务器和负载平衡器使用的东西。
```

> 为了改进安排，Atlas 中的每个 Servlet 都是其自己的集群。默认情况下这个思路隔离了它们，因为行为异常的路由应该只会影响同一 Atlasservlet 中的其他路由。同样，此决定允许独立去推送代码。此外，Dropbox 决定在 [gRPC](https://grpc.io/) 上进行标准化。为了继续处理 HTTP 流量，他们使用了 [GRPC JSON 转码器](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_json_transcoder_filter)中提供了的 [Envoy](https://www.envoyproxy.io/) —— 他们在 Servlet 的前面作为代理服务器和负载平衡器使用的东西。

Author: ```suggestion
为了改进编排，Atlas 中的每个 Servlet 都是其自己的集群。默认情况下该策略提供隔离能力，因为行为异常的路由应该只会影响同一 Atlasservlet 中的其他路由。同样，此决定允许独立去推送代码。此外，Dropbox 决定在 [gRPC](https://grpc.io/) 上进行标准化。为了继续处理 HTTP 流量，他们使用了 [GRPC JSON 转码器](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_json_transcoder_filter)中提供了的 [Envoy](https://www.envoyproxy.io/) —— 他们在 Servlet 的前面作为代理服务器和负载平衡器使用的东西。
```

![HTTP 转码器](https://res.infoq.com/news/2021/03/dropbox-atlas/en/resources/1Dropbox-atlas-http-transcoding-1615307468739.png)

<small>HTTP 转码器，图源 [Dropbox](https://dropbox.tech/infrastructure/atlas--our-journey-from-a-python-monolith-to-a-managed-platform) </small>

> 根据作者的说法，关于可操作性，“Atlas 的秘密秘诀是可管理的经验”。这项工作的主要支柱是自动金丝雀分析，它可以在每个代码推入生产之前自动检查每个代码推入，以及自动缩放功能，从而消除了对容量规划的大部分需求。

Author: ```suggestion
根据他们的说法，在可操作性问题上，“Atlas 的秘密秘诀是可管理的经验”。这项工作的主要依靠是使用自动化运行 Canary 分析，让每个代码推入生产之前自动检查每个代码推入，以及自动拓展功能，从而消除了对容量规划的大部分需求。
```
